docs/proposals/04-nat-gateway.md

```diff
@@ -0,0 +1,105 @@
+
+## Motivation
+Some machines of the network have no public IP addresses. But those machines also need public internet access. A NAT gateway provides this functionality by providing a default gateway to the private machines and an address translation for incoming and outgoing traffic. OnMetal needs also to provide access to private machines (e.g. downloading container images). A standard setup is to use a VirtualIp ([OEP-1](01-networking-integration.md)) but this exposes the machines with a public IP address. This service introduces NAT gateway for everyone not having a public IP address.
+
+### Goals
+- define a user facing API (onMetal api) for the NAT Gateway
+- defining the maximum outgoing connections of any machine in the network
+- defining the (public) IP Address of the gateway
+
+### Non-Goals
+- gateways are network function 
+
+## Proposal
+The NatGateway definition allows to define a gateway for a network. The prefix is implying the Network (VNI) this NAT-Gateway runs on. Better: every machine in that prefix will get a NAT-Gateway. If a machine gets a virtual IP it no longer needs a NAT-Gateway assignment. Anything except the `natIPs` is immutable on the NatGateway. Those are used to scale the size of the Gateway. Changes of the `PortPerMachine` cause the whole NAT-Gateway to be recomputed and can therefore only happen on initializing
+
+### User defined object
+```yaml
+apiVersion: networking.api.onmetal.de/v1alpha1
+kind: NATGateway
+metadata:
+  name: my-nat-4711ab
+  namespace: customer-1
+spec:
+  # the network the NATGateway operates on
+  networkRef: 
+    name: sample-network
+  # public IP addresses the NATGateway has. Every IP has a total of 64512 (655356-1024) ports being available to machines in the NAT domain To have more ports available add more IP addresses 
+  #TODO other types of IP addresses (e.g. internal NATGateway) and pinned IP types
+  type: Public
+  natIPs:
+    - ipFamily: IPv4
+  # defines the concurrent connections per machine and target. 64 is the default (if omitted), must be a power of 2
+  PortsPerMachine: 64                    # 64, 128, 256, 512 or 1024
+  # a selector for the NetworkInterfaces to be part of the NATGateway. That way it is possible to define interfaces that have explicit Internet access and interfaces that do not have. All interfaces are part of NetworkRef, mathing the label by k8s label selector rules and have no VirtualIP 
+  networkInterfaceSelector:
+    matchLabels:
+      key: db
+      foo: bar
+status:
+  ips:
+    - 48.86.152.12    
+  # information of Network interfaces without a virtualIP is needed
+  portsUsed: {{ PortsPerMachine * entries in NATGatewayRouting.destinations }}
+  portsAvailable: {{ ( entries in natIPs * 64512 ) - PortsUsed}}
+```
+
+### internal reconcile object (not customer visible)
```

Comment:
Same as for `LoadBalancerRouting`, customer visibility is not a concern of the API but of a cluster admin defining RBAC on top. We should rephrase the wording the same way we did for `LoadBalancerRouting`.

Reply:
yes fixed